### PR TITLE
[ICDS Dashboard]Awc eligibility for performance report

### PR DESCRIPTION
**Pull request by challabeehyv**
##### SUMMARY
Added eligibility check for performance report

##### FEATURE FLAG
No

### CODE REVIEW

File: custom/icds_reports/tasks.py
Comment (czue) [inline custom/icds_reports/tasks.py]: Casting the querysets to lists:

1. Will cause everything to be loaded from the DB which could be expensive
2. Makes the variable names ending in `_queryset` very misleading.

Why is it necessary and is there any way to push more logic into the DB instead?

File: custom/icds_reports/tasks.py
Comment (czue) [inline custom/icds_reports/tasks.py]: This method name is misleading. Starting with `is_` implies that it would be a boolean check, but if successful it returns a list of four things. I would recommend renaming it to be more clear, e.g. `get_row_data_if_awc_launch_data_is_wrong` (not sure that's right but hopefully the idea is clear)

File: custom/icds_reports/tasks.py
Comment (czue) [inline custom/icds_reports/tasks.py]: Code like this is very difficult to read. If there isn't a way to make it clearer, it would be helpful to have a comment or doc string describing what the expected input format is to help understand the intent of this. 

But probably it'd be better to extract it to a well-named method, e.g. `_performance_row_has_no_conduct_report` (again I am not sure about the names).

This comment applies many places in this PR

**Review by czue — CHANGES_REQUESTED** [review]
Can you respond to the first comments? Then can continue a review once the PR is a bit easier to understand.

File: custom/icds_reports/tasks.py
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: we are using the same records in different checks. So i thought loading them into memory and using them in various checks reduces the db calls. Also we are iterating only once for all the checks and passing the resultset rows as parameters

File: custom/icds_reports/tasks.py
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: ok..will do

File: custom/icds_reports/tasks.py
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: okay

File: custom/icds_reports/tasks.py
Comment (czue) [reply custom/icds_reports/tasks.py]: Ok, well then at a minimum I think change the variable names.

Do you know how many rows are expected to be returned by this? Only reason it could be an issue is if the results are very large.

Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: agreed. we can get them by pagination. Trade off is addition of latency

File: custom/icds_reports/tasks.py
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: changed

File: custom/icds_reports/tasks.py
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: changed

File: custom/icds_reports/tasks.py
Comment (Rohit25negi) [inline custom/icds_reports/tasks.py]: How do we sure that both lists are ordered and rows are index i from both list corresponds to same month and same awc?

@challabeehyv

File: custom/icds_reports/tasks.py
Comment (Rohit25negi) [inline custom/icds_reports/tasks.py]: This is going to be slower and slower by time Because this is comparing data from infinite past. 
It might be better to just compare for the latest month. Whats your thoughts @calellowitz

Comment (Rohit25negi) [reply custom/icds_reports/tasks.py]: Also, There could be lacs of rows and that would increase with the frequency of lacs.

File: custom/icds_reports/tasks.py
Comment (Rohit25negi) [inline custom/icds_reports/tasks.py]: This entire validation code block does not match with the function name. Better to take out this entire validation block into another function. and call that in this function with single line
@challabeehyv

File: custom/icds_reports/tasks.py
Comment (Rohit25negi) [inline custom/icds_reports/tasks.py]: Better to add this info as function doc instead of inline comment.

File: custom/icds_reports/tasks.py
Comment (Rohit25negi) [inline custom/icds_reports/tasks.py]: This has to be filtered out for aggregation_level=5

File: custom/icds_reports/tasks.py
Comment (Rohit25negi) [inline custom/icds_reports/tasks.py]: It looks first value in the row is just the awc_id, then  no need to include both `performance_row[0]` and  `awcagg_row[0]`.

File: custom/icds_reports/tasks.py
Comment (Rohit25negi) [inline custom/icds_reports/tasks.py]: Why do we need different emails for different checks? Can we send one email showing result of all the checks of performance report? I dont want a check to get missed if one of the email fails to deliver.

**Review by Rohit25negi — CHANGES_REQUESTED** [review]
Left a bunch of comments. Some code correction and some code cleaning. 
PS: hold of the "latest month comparison" one comment until @calellowitz  gives his thought. But please address others. Feel free to ping in case of query.

File: custom/icds_reports/tasks.py
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: since we are filtering with valid awc_ids, i think this can be left out

File: custom/icds_reports/tasks.py
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: as per @calellowitz , they should be same

File: custom/icds_reports/tasks.py
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: okay..i will change it

File: custom/icds_reports/tasks.py
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: okay

File: custom/icds_reports/tasks.py
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: since the checks are relying on the order of awc_ids, i thought to add ids from both just in case if there is a mismatch, we can eliminate this dilemma

File: custom/icds_reports/tasks.py
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: I thought there will be cases where receipients or content will be different for different cases.If not, we can merge and send them as a single mail

File: custom/icds_reports/tasks.py
Comment (calellowitz) [inline custom/icds_reports/tasks.py]: These checks are very repetitive, it would be nice to pull this out into a function

File: custom/icds_reports/tasks.py
Comment (calellowitz) [inline custom/icds_reports/tasks.py]: These functions are all almost identical (except the columns), they could probably be combined into one helper that takes columns as an argument

File: custom/icds_reports/tasks.py
Comment (calellowitz) [inline custom/icds_reports/tasks.py]: This will break if one includes a row not in the other, which i believe should not happen, but it would be good to guard against.

File: custom/icds_reports/tasks.py
Comment (calellowitz) [inline custom/icds_reports/tasks.py]: These might be clearer if you use `values` instead of `values_list` because then the dict keys will explain what you are accessing.

File: custom/icds_reports/tasks.py
Comment (calellowitz) [inline custom/icds_reports/tasks.py]: If valid_visits is 0, they are not eligible unless visit denominator is also 0.
ie (0/0 is eligible but 0/4 is not)

File: custom/icds_reports/tasks.py
Comment (calellowitz) [inline custom/icds_reports/tasks.py]: same comment here as above 0/0 counts. 0/5 does not

File: custom/icds_reports/tasks.py
Comment (calellowitz) [inline custom/icds_reports/tasks.py]: an awc is also not eligible if it is not launched

**Review by calellowitz — CHANGES_REQUESTED** [review]
This looks much better than the last PR but a few more comments and will let cory take another look as well

File: custom/icds_reports/tasks.py
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: if we observe a mismatch should we skip those rows.But the subsequent rows will also be skipped. I see a lot of data leak. instead we can merge the rows using dataframes and filter the mismatch rows beforehand and proceed the checks with awc_id macthed rows. Is it an overkill?

File: custom/icds_reports/tasks.py
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: got it

File: custom/icds_reports/tasks.py
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: covered in this case

File: custom/icds_reports/tasks.py
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: got it

**Review by calellowitz — CHANGES_REQUESTED** [review]
Few changes left it looks like. I think if you could also add pagination to the db queries that would be great. I think it will be fine for now. But these will still return a large number of results and that number will only increase over time.

File: custom/icds_reports/tasks.py
Comment (calellowitz) [inline custom/icds_reports/tasks.py]: This should be logged as missing as well. Just like missing rows in the performance report.

File: custom/icds_reports/tasks.py
Comment (calellowitz) [inline custom/icds_reports/tasks.py]: These need to be filtered by date as well

**Review by calellowitz — CHANGES_REQUESTED** [review]
Looks like you are almost there. Few small changes left

File: custom/icds_reports/tasks.py
Comment (challabeehyv) [reply custom/icds_reports/tasks.py]: added the filter

File: custom/icds_reports/tasks.py
Comment (calellowitz) [inline custom/icds_reports/tasks.py]: not blocking but these don't need to be coerced to a list. you should be able to just iterate right over them, and it is probably more efficient that way

**Review by calellowitz — APPROVED** [review]
This is good enough for me now. Some non blocking comments and will wait for approval from the other two but i am comfortable merging

**Review by calellowitz — COMMENTED** [review]
I know I said, i was good to merge, but can you actually make this a celery task so it happens asynchronously. just need to add an `@task(queue='icds_dashboard_reports_queue', serializer='pickle')` decorator, and then call it as `aggregate_validation_helper.delay(agg_date)`

**Comment by challabeehyv** [comment]
> @task(queue='icds_dashboard_reports_queue', serializer='pickle')

added

**Comment by calellowitz** [comment]
@czue and @Rohit25negi 
Are you planning on doing another review or should I just merge

**Comment by czue** [comment]
sorry, i was not planning to review after you did unless you specifically wanted me to @calellowitz

**Review by Rohit25negi — APPROVED** [review]
Looks good to me. Ready for merge  once merge conflict is resolved. @challabeehyv 

can you also add a bit more context on the PR description or atleast add the ticket link in the description to make description more contextual.

File: custom/icds_reports/tasks.py
Comment (calellowitz) [reply custom/icds_reports/tasks.py]: I think I might be missing something in the comment. This is being called async in a separate celery task so that it wont block the generation. At this point, that is to prevent a false positive from breaking the report in addition to the point you made about not wanting things like email failure to be blocking.

Comment (Rohit25negi) [reply custom/icds_reports/tasks.py]: hey, I made this comment on the initial stage of the PR when all the checking was happening with sequential manner in same function. Now that after its made async, so, all good now for me.

**Pull request merged**
